apps/aecore/src/aec_upnp.erl

```diff
@@ -31,17 +32,34 @@
 %% to 7200 seconds (two hours). No recommendation for UPnP found.
 -define(MAPPING_LIFETIME, 7200).
 
+-record(state, {pid :: pid() | undefined,
+                mon :: reference() | undefined}).
+
 %%%===================================================================
 %%% API
 %%%===================================================================
 
+-spec is_enabled() -> boolean().
 is_enabled() ->
     case aeu_env:find_config([<<"sync">>, <<"upnp_enabled">>],
                              [user_config, schema_default]) of
         {ok, IsEnabled} -> IsEnabled;
         undefined       -> false
     end.
 
+-spec prep_stop() -> ok.
+prep_stop() ->
+    case is_enabled() of
+        true ->
+            case catch gen_server:call(?MODULE, prep_stop, 10000) of
```

Comment:
It seems you rely on `aecore_app:prep_stop` rather than in `terminate` of `aec_upnp` in order to increase the 5000 ms the `aecore_sup` gives this child? (It is tricky as this worker process is now acting partially as a supervisor, and the timeout for a supervisor is recommended to be infinity in order not to kill the supervisor before it had killed the unliked children.)
Ok.

Reply:
Prep_stop is called before application shutting down process begin. Then the entire supervision tree of the application, including included applications, is terminated in reversed start order, so we should be safe to keep the mapping cleanup in `terminate/2`. But it seems to be safer/clearer anyway to clean up mapping before shutting down procedure begins (cleanup procedure rely on other applications, e.g. `inets`, to be started).

Infinity is not used in case some nat discovery/mapping removal hangs in the lib, but should never happen (I used some artificial sleeps in nat lib to provoke very unlikely scenarios).